calicoctl/commands/ipam/show.go

```diff
@@ -24,80 +24,136 @@ import (
 	"github.com/olekukonko/tablewriter"
 
 	"github.com/projectcalico/calicoctl/calicoctl/commands/constants"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/clientv3"
 	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
 	"github.com/projectcalico/libcalico-go/lib/ipam"
 
 	docopt "github.com/docopt/docopt-go"
 
 	"github.com/projectcalico/calicoctl/calicoctl/commands/argutils"
 	"github.com/projectcalico/calicoctl/calicoctl/commands/clientmgr"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
 )
 
-// IPAM takes keyword with an IP address then calls the subcommands.
-func Show(args []string) error {
-	doc := constants.DatastoreIntro + `Usage:
-  calicoctl ipam show [--ip=<IP> | --show-blocks] [--config=<CONFIG>]
+type borrowedIP struct {
+	addr           string
+	borrowingNode  string
+	block          string
+	blockOwner     string
+	allocationType string
+	allocatedTo    string
+}
 
-Options:
-  -h --help             Show this screen.
-     --ip=<IP>          Report whether this specific IP address is in use.
-     --show-blocks      Show detailed information for IP blocks as well as pools.
-  -c --config=<CONFIG>  Path to the file containing connection configuration in
-                        YAML or JSON format.
-                        [default: ` + constants.DefaultConfigPath + `]
+func getPoolDetails(ctx context.Context, ippoolClient clientv3.IPPoolInterface, bc bapi.Client) ([]*borrowedIP, error) {
+	var details []*borrowedIP
 
-Description:
-  The ipam show command prints information about a given IP address, or about
-  overall IP usage.
-`
-	parsedArgs, err := docopt.Parse(doc, args, true, "", false, false)
+	// Read details for all blocks.
+	blocks, err := bc.List(ctx, model.BlockListOptions{}, "")
 	if err != nil {
-		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
-	}
-	if len(parsedArgs) == 0 {
-		return nil
+		return nil, err
 	}
+	for _, kvp := range blocks.KVPairs {
+		b := kvp.Value.(*model.AllocationBlock)
 
-	ctx := context.Background()
+		for i := range b.Allocations {
+			if b.Allocations[i] != nil {
+				// Allocations[i] if not nil:
+				// - b.OrdinalToIP(i) is the allocated IP address
+				// - *b.Allocations[i] is the index of the corresponding b.Attributes
 
-	// Create a new backend client from env vars.
-	cf := parsedArgs["--config"].(string)
-	client, err := clientmgr.NewClient(cf)
+				attributes := b.Attributes[*b.Allocations[i]]
+
+				if b.Affinity != nil {
```

Comment:
Do we want to show IP addresses that are allocated in blocks that have no affinity? That is potentially also considered a form of borrowing.

If we _do_ want to show that, then we'll need to restructure this a little bit.

Reply:
Wouldn't that be that too many on large cluster?

Reply:
We don't expect an exceptionally large number of these. At least not any different than we expect for other borrowed addresses.

Fundamentally, on a large cluster, there might be a lot of IP address allocations! But I think we should probably show these since most likely the reason they are there is because they were borrowed originally.

Reply:
Yeah I had misunderstood your feedback. Then after re-reading it I had deleted my question :-)